src/Domain/Services/DraftOrders.php

```diff
@@ -0,0 +1,151 @@
+			add_filter( 'woocommerce_register_shop_order_post_statuses', [ __CLASS__, 'register_draft_order_post_status' ] );
+			add_filter( 'woocommerce_valid_order_statuses_for_payment', [ __CLASS__, 'append_draft_order_post_status' ] );
+			add_action( 'woocommerce_cleanup_draft_orders', [ __CLASS__, 'delete_expired_draft_orders' ] );
+			add_action( 'admin_init', [ $this, 'install' ] );
+		} else {
+			// Maybe remove existing cronjob if present because it shouldn't be needed in the environment.
+			$this->maybe_remove_cronjobs();
+		}
+	}
+
+	/**
+	 * Installation related logic for Draft order functionality.
+	 */
+	public function install() {
+		$this->maybe_create_cronjobs();
+	}
+
+
+	/**
+	 * Remove cronjobs if they exist (but only from admin).
+	 */
+	protected function maybe_remove_cronjobs() {
+		if ( ! is_admin() ) {
+			return;
+		}
+		if ( function_exists( 'as_next_scheduled_action' ) && true === as_next_scheduled_action( 'woocommerce_cleanup_draft_orders' ) ) {
+			as_unschedule_all_actions( 'woocommerce_cleanup_draft_orders' );
+		}
+	}
+
+	/**
+	 * Maybe create cron events.
+	 */
+	protected function maybe_create_cronjobs() {
+		if ( function_exists( 'as_next_scheduled_action' ) && false === as_next_scheduled_action( 'woocommerce_cleanup_draft_orders' ) ) {
+			as_schedule_recurring_action( strtotime( 'midnight tonight' ), DAY_IN_SECONDS, 'woocommerce_cleanup_draft_orders' );
+		}
+	}
+
+
+	/**
+	 * Register custom order status for orders created via the API during checkout.
+	 *
+	 * Draft order status is used before payment is attempted, during checkout, when a cart is converted to an order.
+	 *
+	 * @param array $statuses Array of statuses.
+	 * @return array
+	 */
+	public static function register_draft_order_status( array $statuses ) {
+		$statuses['wc-checkout-draft'] = _x( 'Draft', 'Order status', 'woo-gutenberg-products-block' );
+		return $statuses;
+	}
+
+	/**
+	 * Register custom order post status for orders created via the API during checkout.
+	 *
+	 * @param array $statuses Array of statuses.
+	 * @return array
+	 */
+	public static function register_draft_order_post_status( array $statuses ) {
+		$statuses['wc-checkout-draft'] = [
+			'label'                     => _x( 'Draft', 'Order status', 'woo-gutenberg-products-block' ),
+			'public'                    => false,
+			'exclude_from_search'       => false,
+			'show_in_admin_all_list'    => false,
+			'show_in_admin_status_list' => true,
+			/* translators: %s: number of orders */
+			'label_count'               => _n_noop( 'Drafts <span class="count">(%s)</span>', 'Drafts <span class="count">(%s)</span>', 'woo-gutenberg-products-block' ),
+		];
+		return $statuses;
+	}
+
+	/**
+	 * Append draft status to a list of statuses.
+	 *
+	 * @param array $statuses Array of statuses.
+	 * @return array
+	 */
+	public static function append_draft_order_post_status( $statuses ) {
+		$statuses[] = 'checkout-draft';
```

Comment:
This was a straight copy/paste from pre-existing code, but I just noticed and wondered, should this be `wc-checkout-draft` @mikejolley?

Reply:
Core expects unprefixed names in this filter.